src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +318,101 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeout q = do
+  timedOutVar ← newEmptyTMVarIO
+
+  let
+    timeoutAction = do
+      threadDelay timeout
+      atomically $ putTMVar timedOutVar ()
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = do
+          res ← Left <$> readTBMQueue q <|> Right <$> readTMVar timedOutVar
+          case res of
+            Left mx → maybe (return xs) (go ∘ (:xs)) mx
+            Right () → return xs
+
+  withAsync timeoutAction $ \h → do
+    xs ← atomically $ go []
+    xs <$ cancel h
```

Comment:
Actually, after thinking a bit about what it would mean to put it into a single `atomically`, I wonder if the current code is even correct as-is.

The thinking was that the `readTMVar` would retry until it the timeout fired, but this would actually end up retrying the entire transaction as a whole, so that doesn't seem right. Do you have any thoughts or clarifications?